python/ray/data/_internal/execution/legacy_compat.py

```diff
@@ -88,7 +88,11 @@ def execute_to_legacy_block_list(
         The output as a legacy block list.
     """
     if DatasetContext.get_current().optimizer_enabled:
-        dag, stats = get_execution_plan(plan._logical_plan).dag, None
+        dag = get_execution_plan(plan._logical_plan).dag
```

Comment:
we should add same logic for `execute_to_legacy_block_iterator` method above as well.

Reply:
Nit: Now that this logic is both here and `execute_to_legacy_block_iterator` (and soon to be another method once Eric's split PR is merged), it might be worth consolidating this logic into a small helper function.

```python
    stats = _get_initial_stats_from_plan(plan)

# ...

def _get_initial_stats_from_plan(plan: ExecutionPlan) -> DatasetStats:
    assert DatasetContext.get_current().optimizer_enabled
    if plan._snapshot_blocks is not None and not plan._snapshot_blocks.is_cleared():
        return plan._snapshot_stats
    else:
        return plan._in_stats
``` 

Or you could just use `ExecutionPlan._get_source_blocks_and_stages()` and only keep the stats, so we really consolidate this logic, but we'd need to make sure that the snapshot isn't cleared on this call.

Reply:
@clarkzinzow i chose to go with consolidating into a `_get_initial_stats_from_plan` method as you suggested above, but if desired, i can go ahead and modify `ExecutionPlan._get_source_blocks_and_stages()` to take an optional param which controls the `self._clear_snapshot()` call inside the method.